signac/core/syncedattrdict.py

```diff
@@ -193,18 +177,28 @@ def clear(self):
         self._data = {}
         self.sync()
 
-    def update(self, mapping):
+    def update(self, other=None, **kwargs):
+        if other is not None:
+            # Convert sequence of key, value pairs to dict before validation
+            if not isinstance(other, Mapping):
+                other = dict(other)
+            self._validate(other)
+        if kwargs:
+            self._validate(kwargs)
         self.load()
         with self._suspend_sync():
-            for key, value in mapping.items():
-                self._data[self._validate_key(key)] = self.from_base(data=value, parent=self)
+            if other:
+                for key, value in other.items():
+                    self._data[key] = self.from_base(data=value, parent=self)
+            for key, value in kwargs.items():
+                self._data[key] = self.from_base(data=value, parent=self)
         self.sync()
 
     def setdefault(self, key, default=None):
+        self._validate({key: default})
         self.load()
         with self._suspend_sync():
-            ret = self._data.setdefault(self._validate_key(key),
-                                        self.from_base(data=default, parent=self))
+            ret = self._data.setdefault(key, self.from_base(default, parent=self))
         self.sync()
         return ret
```

Comment:
While it is possibly premature, I think it's worth applying some minor easy optimizations to `setdefault`. If the key is already in the dict, there's no need to call `_validate`, `_data._setdefault`, or `from_base`: you can just set `ret = self._data[key]` after a load and skip the sync. I can't directly suggest because there are deleted lines in this function, but I'm thinking something like:
```
def setdefault(self, key, default=None):
    self.load()
    try:
        ret = self._data[key]
    except KeyError:
        self._validate({key: default})
        ret = self.from_base(default, parent=self))
        with self._suspend_sync():
            self._data[key] = ret
        self.sync()
    return ret
```
This method avoids a lot of extra work when the key is in the dict. I purposely chose not to use `self.__getitem__(key)` to avoid depending on that function calling `load` (in case we somehow change that implementation, just trying to avoid subtle bugs). I didn't test this implementation, please make sure I didn't make any errors, in particular with the order of when I call `from_base`.

Reply:
@vishav1771 @vyasr Be aware that handling exceptions can be expensive. I recommend LBYL (`if key in self._data:` / `else:`) instead of EAFP (`try:` / `except KeyError:`) here.

Update code sample:
```python
def setdefault(self, key, default=None):
    self.load()
    if key in self._data:
        ret = self._data[key]
    else:
        self._validate({key: default})
        ret = self.from_base(default, parent=self))
        with self._suspend_sync():
            self._data[key] = ret
        self.sync()
    return ret
```

Evidence / demo of LBYL being faster:

```python
import timeit

hit = """data = {'foo': 0, 'bar': 0}"""
miss = """data = {'baz': 0, 'bar': 0}"""

lbyl = """if 'foo' in data:
  data['foo'] += 1
else:
  data['bar'] += 1
"""

eafp = """try:
  data['foo'] += 1
except KeyError:
  data['bar'] += 1
"""

hit_lbyl_time = timeit.timeit(setup=hit, stmt=lbyl, number=1000000)
hit_eafp_time = timeit.timeit(setup=hit, stmt=eafp, number=1000000)
miss_lbyl_time = timeit.timeit(setup=miss, stmt=lbyl, number=1000000)
miss_eafp_time = timeit.timeit(setup=miss, stmt=eafp, number=1000000)

print('Hit, LBYL:', hit_lbyl_time)
print('Hit, EAFP:', hit_eafp_time)
print('Miss, LBYL:', miss_lbyl_time)
print('Miss, EAFP:', miss_eafp_time)
```

Results:
```
Hit, LBYL: 0.0599388
Hit, EAFP: 0.0484887
Miss, LBYL: 0.0627916
Miss, EAFP: 0.2142205
```

Reply:
That's true, but the typical use case for `setdefault` is when keys are likely to be present. In this case EAFP expresses the expected usage pattern more appropriately. Having said that, I don't think there's any "representative" use case to know how the trade offs play out in practice. Based on your timings you would need >10:1 hit:miss ratio for EAFP to be faster. That's actually probably a pretty realistic in any case where I would actually use `setdefault`; in practice unless I expect mostly hits I write LBYL code myself rather than use `setdefault`. However, I don't really know what others usually do (tbh in my experience very few people use this method in general) so I don't have a strong opinion if your usage pattern indicates that switching would constitute a meaningful performance improvement.

Reply:
I would find it hard to justify an assumed ratio of >10:1 or more. I lean towards LBYL, because we cannot know the true ratio to expect. For example, I have used `setdefault` in cases where I end up using the “default” more frequently than the stored value, which is the opposite of the “mostly read from existing key” (mostly hits) scenario posed above.

Reply:
I looked through the CPython standard library and it’s a mixture of LBYL and EAFP approaches, depending on the class. There is no clear preference as far as I can tell.

Reply:
That's fine with me. Frankly I almost always end up using a `defaultdict` anywhere that I might otherwise use `setdefault`, the only real exception being if I'm modifying a dict produced by some third party code, so I'm probably not a representative user of `setdefault` insofar as such a user still exists in Python 3.8.